controllers/appwrapper_controller.go

```diff
@@ -414,35 +439,102 @@ func IsAwPending() (false bool, aw *arbv1.AppWrapper) {
+	if err != nil {
+		klog.Infof("Error deleted label to machines %v", err)
+	}
+}
+
+func addLabelToMachine(aw *arbv1.AppWrapper, machineName string) {
+	labelPatch := fmt.Sprintf(`[{"op":"add","path":"/metadata/labels/%s","value":"%s" }]`, aw.Name, aw.Name)
+	_, err := machineClient.MachineV1beta1().Machines(namespaceToList).Patch(context.Background(), machineName, types.JSONPatchType, []byte(labelPatch), metav1.PatchOptions{})
+	if err != nil {
+		klog.Infof("Error adding label to machines %v", err)
+	}
+}
+
+func addLabelToNode(aw *arbv1.AppWrapper, nodeName string) {
+	labelPatch := fmt.Sprintf(`[{"op":"add","path":"/metadata/labels/%s","value":"%s" }]`, aw.Name, aw.Name)
+	_, err := kubeClient.CoreV1().Nodes().Patch(context.Background(), nodeName, types.JSONPatchType, []byte(labelPatch), metav1.PatchOptions{})
+	if err != nil {
+		klog.Infof("Error adding label to machine %v", err)
+	}
+}
+
+// add logic to swap out labels with new appwrapper label
+func checkIfExactMatchExists(aw *arbv1.AppWrapper) bool {
+	var shouldScaleDown bool = true
+	allAw, err := queueJobLister.List(labels.Everything())
+	if err != nil {
+		klog.Error("Cannot list queued appwrappers, associated machines will be deleted")
+		return shouldScaleDown
+	}
+	var existingAcquiredMachineTypes = ""
+
+	for key, value := range aw.Labels {
+		if key == "orderedinstance" {
+			existingAcquiredMachineTypes = value
+		}
+	}
+
+	for _, eachAw := range allAw {
+		for k, v := range eachAw.Labels {
+			if k == "orderedinstance" {
+				if v == existingAcquiredMachineTypes {
+					shouldScaleDown = false
+					klog.Infof("Found exact match, %v appwrapper has acquire machinetypes %v", aw.Name, existingAcquiredMachineTypes)
+					allMachines, errm := machineClient.MachineV1beta1().Machines(namespaceToList).List(context.Background(), metav1.ListOptions{})
+					if errm != nil {
+						klog.Infof("Error creating machineset: %v", errm)
+					}
+					//klog.Infof("Got all machines %v", allMachines)
+					for idx := range allMachines.Items {
+						machine := &allMachines.Items[idx]
+						for k, _ := range machine.Labels {
+							if k == aw.Name {
+								nodeName := machine.Status.NodeRef.Name
+								klog.Infof("removing label from node %v that belonged to appwrapper %v and adding node to new appwrapper %v", nodeName, aw.Name, eachAw.Name)
+								removeLabelFromMachine(aw, machine.Name)
+								removeLabelFromNode(aw, nodeName)
+								addLabelToMachine(eachAw, machine.Name)
+								addLabelToNode(eachAw, nodeName)
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+	return shouldScaleDown
+
+}
+
 func onDelete(obj interface{}) {
 	aw, ok := obj.(*arbv1.AppWrapper)
 	if ok {
 		klog.Infof("Appwrapper deleted scale-down machineset: %s ", aw.Name)
-		// if reUseMachineSet() {
-		// 	return
-		// }
-		scaleDown(aw)
+		if reuse {
+			if checkIfExactMatchExists(aw) {
```

Comment:
Why does `checkIfExactMatchExists()` return a bool "shouldScaleDown" if either way `scaleDown()` is going to be called the same way? It seems like the "check" part of the function is unused, and it is more so just a means of doing label swapping if a match is found. Also the conditional here seems unnecessary. And for clarification, what exactly is the goal of "annotateToDeleteMachine"? I'm assuming it is to delete all machines for the scaling down AW that didn't have their labels swapped to a new pending AW? I could take a stab at restructuring/simplifying this scaling down portion if you'd like.

Reply:
Good points, If the exact match does not exist, we want to scale down. Machines in Machinesets need to be cherry-picked to be deleted.
Can you create an issue and add the restructuring you are planning to add?